osu.Game/Rulesets/UI/GameplaySampleTriggerSource.cs

```diff
@@ -68,27 +69,61 @@ public virtual void Play()
         protected HitObject GetMostValidObject()
         {
             // The most optimal lookup case we have is when an object is alive. There are usually very few alive objects so there's no drawbacks in attempting this lookup each time.
-            var hitObject = hitObjectContainer.AliveObjects.FirstOrDefault(h => h.Result?.HasResult != true)?.HitObject;
+            var drawableHitObject = hitObjectContainer.AliveObjects.FirstOrDefault(h => h.Result?.HasResult != true);
+
+            if (drawableHitObject != null)
+            {
+                // A hit object may have a more valid nested object.
+                drawableHitObject = getMostValidNestedDrawable(drawableHitObject);
+
+                return drawableHitObject.HitObject;
+            }
 
             // In the case a next object isn't available in drawable form, we need to do a somewhat expensive traversal to get a valid sound to play.
-            if (hitObject == null)
+            // This lookup can be skipped if the last entry is still valid (in the future and not yet hit).
+            if (fallbackObject == null || fallbackObject.Result?.HasResult == true)
             {
-                // This lookup can be skipped if the last entry is still valid (in the future and not yet hit).
-                if (fallbackObject == null || fallbackObject.Result?.HasResult == true)
-                {
-                    // We need to use lifetime entries to find the next object (we can't just use `hitObjectContainer.Objects` due to pooling - it may even be empty).
-                    // If required, we can make this lookup more efficient by adding support to get next-future-entry in LifetimeEntryManager.
-                    fallbackObject = hitObjectContainer.Entries
-                                                       .Where(e => e.Result?.HasResult != true).MinBy(e => e.HitObject.StartTime);
-
-                    // In the case there are no unjudged objects, the last hit object should be used instead.
-                    fallbackObject ??= hitObjectContainer.Entries.LastOrDefault();
-                }
-
-                hitObject = fallbackObject?.HitObject;
+                // We need to use lifetime entries to find the next object (we can't just use `hitObjectContainer.Objects` due to pooling - it may even be empty).
+                // If required, we can make this lookup more efficient by adding support to get next-future-entry in LifetimeEntryManager.
+                fallbackObject = hitObjectContainer.Entries
+                                                   .Where(e => e.Result?.HasResult != true).MinBy(e => e.HitObject.StartTime);
+
+                if (fallbackObject != null)
+                    return fallbackObject.HitObject;
```

Comment:
This early return is suspicious to me. It is the only one out of three return pathways which consider an object without a result, that does not call `getEarliestNestedObject()` on it prior to return. (I am not counting the one return pathway that considers an object which *has* a result here.)

I'd expect this early-return to either not be here at all, or return `getEarliestNestedObject(fallbackObject.HitObject)` instead. Am I missing something?

(I tried to write a test to try and break this but am unable to get a failing one after trying a little while. So this is mostly just a hunch based off of exit path analysis.)

Reply:
I think you're right. I wrote this before realising we wanted the earliest nested in cases like this, and probably did not update this path.

Have changed to use `getEarliestNestedObject()`.